tests/integration_test.rs

```diff
@@ -2,49 +2,126 @@
+                      password: String)
+                      -> (hyper::status::StatusCode, String) {
         let encoded = form_urlencoded::Serializer::new(String::new())
             .append_pair("username", uname.as_ref())
             .append_pair("token", token.as_ref())
             .append_pair("password", password.as_ref())
             .finish();
-        let res = c.post("http://localhost:6767").body(encoded.as_bytes()).send();
+        let mut res = c.post("http://localhost:6767")
+            .header(hyper::header::ContentType::form_url_encoded())
+            .body(encoded.as_bytes())
+            .send()
+            .unwrap();
+        let mut body: String = "".to_string();
+        res.read_to_string(&mut body).unwrap();
+
+        (res.status, body)
+    }
+
+    // TODO make an externally usable matrix client to handle this stuff
+    struct PasswordLoginRequest {
+        user: String,
+        password: String,
+    }
 
-        res.unwrap().status
+    impl ToJson for PasswordLoginRequest {
+        fn to_json(&self) -> Json {
+            let mut obj = BTreeMap::new();
+            // https://matrix.org/docs/spec/client_server/r0.2.0.html#login
+            obj.insert("type".to_string(), "m.login.password".to_json());
+            obj.insert("user".to_string(), self.user.to_json());
+            obj.insert("password".to_string(), self.password.to_json());
+            Json::Object(obj)
+        }
     }
 
-#[test]
+    fn login_with_user(c: &hyper::Client,
+                       login: PasswordLoginRequest)
+                       -> Result<(), hyper::status::StatusCode> {
+        let resp = c.post("http://localhost:8080/_matrix/client/r0/login")
+            .header(hyper::header::ContentType::json())
+            .body(format!("{}", login.to_json()).as_bytes())
+            .send().unwrap();
+
+        match resp.status {
+            hyper::status::StatusCode::Ok=> Ok(()),
+            _ => Err(resp.status),
+        }
+    }
+
+    #[test]
     fn it_works() {
         let c = Client::new();
-        let orig_pass = "password1234".to_string();
+        let mut rng = OsRng::new().unwrap();
+        let orig_pass = format!("password{}", rng.next_u64());
+        let username1 = format!("user{}", rng.next_u64());
 
-        register_user(&c, "bob".to_string(), orig_pass.clone());
+        register_user(&c, username1.clone(), orig_pass.clone());
+        let result = login_with_user(&c,
+                                     PasswordLoginRequest {
+                                         user: username1.clone(),
+                                         password: orig_pass.clone(),
+                                     });
+        assert!(result.is_ok());
 
-        // TODO: login with the above username and password
+        let token = format!("token{}", rng.next_u64());
+        let full_uname = format!("@{}:synapse_password_reset.local", username1);
+        let new_pass = format!("newpassword{}", rng.next_u64());
```

Comment:
Intentionally dropped because the function that truncates and handles consistently is now bcrypt::hash, and I added a test for that in my 'the longest of passwords' branch